service/history/replication/ack_manager.go

```diff
@@ -123,32 +124,42 @@ func (p *ackMgrImpl) NotifyNewTasks(
 		return
 	}
 	maxTaskID := tasks[0].GetTaskID()
+	maxVisibilityTimestamp := tasks[0].GetVisibilityTime()
 	for _, task := range tasks {
 		if maxTaskID < task.GetTaskID() {
 			maxTaskID = task.GetTaskID()
 		}
+		if maxVisibilityTimestamp.Before(task.GetVisibilityTime()) {
+			maxVisibilityTimestamp = task.GetVisibilityTime()
+		}
 	}
 
 	p.Lock()
 	defer p.Unlock()
 	if p.maxTaskID == nil || *p.maxTaskID < maxTaskID {
 		p.maxTaskID = &maxTaskID
 	}
+	if p.maxTaskVisibilityTimestamp == nil || p.maxTaskVisibilityTimestamp.Before(maxVisibilityTimestamp) {
+		p.maxTaskVisibilityTimestamp = timestamp.TimePtr(maxVisibilityTimestamp)
+	}
 }
 
-func (p *ackMgrImpl) GetMaxTaskID() int64 {
+func (p *ackMgrImpl) GetMaxTaskInfo() (int64, time.Time) {
 	p.Lock()
 	defer p.Unlock()
 
-	if p.maxTaskID == nil {
+	// MaxTaskID and MaxTaskVisibilityTimestamp should be set in the same place.
+	if p.maxTaskID == nil || p.maxTaskVisibilityTimestamp == nil {
 		// maxTaskID is nil before any replication task is written which happens right after shard reload. In that case,
 		// use ImmediateTaskMaxReadLevel which is the max task id of any immediate task queues.
 		// ImmediateTaskMaxReadLevel will be the lower bound of new range_id if shard reload. Remote cluster will quickly (in
 		// a few seconds) ack to the latest ImmediateTaskMaxReadLevel if there is no replication tasks at all.
-		return p.shard.GetQueueExclusiveHighReadWatermark(tasks.CategoryReplication, p.currentClusterName).Prev().TaskID
+		maxTaskID := p.shard.GetQueueExclusiveHighReadWatermark(tasks.CategoryReplication, p.currentClusterName).Prev().TaskID
+		maxVisibilityStampe := p.shard.GetTimeSource().Now()
+		return maxTaskID, maxVisibilityStampe
```

Comment:
Not sure if it is possible that `p.maxTaskID` is nil while `p.maxTaskVisibilityTimestamp` (and inverse case too) is not but current code will skip non nil value. Shall we instead

```
maxTaskID := p.maxTaskID
if maxTaskID == nil {
    maxTaskID =  p.shard.GetQueueExclusiveHighReadWatermark(tasks.CategoryReplication, p.currentClusterName).Prev().TaskID
}
```
same for `maxVisibilityStamp`

Reply:
Updated